plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Scoping.hs

```diff
@@ -0,0 +1,282 @@
+        TermBind ann strictness varDecl $ referenceVia reg name term
+    referenceVia _ _ typeBind@TypeBind{} = typeBind
+    referenceVia _ _ datatypeBind@DatatypeBind{} = datatypeBind
+
+instance Reference tyname t => Reference (TyVarDecl tyname ann) t where
+    referenceVia reg = referenceVia reg . tyVarDeclName
+
+instance Reference name t => Reference (VarDecl tyname name uni fun ann) t where
+    referenceVia reg = referenceVia reg . varDeclName
+
+instance (Reference TyName t, Reference Name t) => Reference (Datatype TyName Name uni fun ann) t where
+    referenceVia reg (Datatype _ dataDecl params matchName constrs)
+        = referenceVia reg dataDecl
+        -- Parameters of a data type are not visible outside of the data type no matter what.
+        . referenceOutOfScope params
+        . referenceVia reg matchName
+        . referenceVia reg constrs
+
+instance (Reference TyName t, Reference Name t) => Reference (Binding TyName Name uni fun ann) t where
+    referenceVia reg (TermBind _ _ varDecl _)  = referenceVia reg varDecl
+    referenceVia reg (TypeBind _ tyVarDecl _)  = referenceVia reg tyVarDecl
+    referenceVia reg (DatatypeBind _ datatype) = referenceVia reg datatype
+
+-- | Establish scoping for each of the parameters by only annotating every parameter with
+-- 'introduceBound'.
+establishScopingParams :: [TyVarDecl TyName ann] -> Quote [TyVarDecl TyName NameAnn]
+establishScopingParams =
+    traverse $ \(TyVarDecl _ paramNameDup paramKind) -> do
+        paramName <- freshenTyName paramNameDup
+        TyVarDecl (introduceBound paramName) paramName <$> establishScoping paramKind
+
+-- See Note [Weird IR data types].
+-- | Establish scoping for the type of a constructor. After mangling the updated constructor expects
+-- an argument of the \"the data type applied to all its parameters\" type, plus the final head
+-- gets applied to the same thing as well (because it's legal to generate a data type whose
+-- constructors do not return the data type applied to its parameters as program generation can be
+-- entirely arbitrary (no pun intended). All this means that mangling turns well-formed programs
+-- into weirdly shaped mess, but we only care about testing scoping here, so that's fine.
+-- Whether the name of the data type is referenced as in-scope or out-of-scope one in the types of
+-- arguments of constructors is controlled by the first argument, which ultimately depends on the
+-- recursivity of the data type.
+establishScopingConstrTy
+    :: (TyName -> NameAnn)
+    -> TyName
+    -> [TyVarDecl TyName NameAnn]
+    -> Type TyName uni ann
+    -> Quote (Type TyName uni NameAnn)
+establishScopingConstrTy regSelf dataName params = goSpine where
+    toDataAppliedToParams reg
+        = mkIterTyApp NotAName (TyVar (reg dataName) dataName)
+        $ map (\(TyVarDecl _ name _) -> TyVar (registerBound name) name) params
+
+    goSpine (TyForall _ nameDup kindDup ty) = do
```

Comment:
They're all different:

1. `renameSpineM` is monadic, updates names, does not change the spine and throws on the wrong final head
2. `establishScopingConstrTy.goSpine` is monadic, does fancy stuff with the `TyForall binding`, changes the spine by adding a bunch of `TyFun`s (and not only in the `TyForall` case, but also in the `TyFun` one), drops the originally generated final head on the floor and makes it the right one
3. the one in `Reference TyName` is pure, does not do anything fancy with `TyForall`, but does change the spine in the `Fun` case and does not drop the final head on the floor

I would much rather not try to unify that. But I do agree it sucks that we need to duplicate this spine-traversing logic with no way to ensure that "spine" means the same thing in different places. But that really should be controlled by types of constructors being special things rather than general `Type`s.

Reply:
Fair enough.